src/Analysis.cpp

```diff
@@ -252,60 +259,132 @@ void CalledMethodsAnalysis::doAnalysis(Function &F,
-      }
+  for (auto Pair : MemoryFunctions) {
+    if (fnName == Pair.first) {
+      errs() << "**ANALYSIS-WARNING**: Re-definition of allocation function '"
+             << fnName
+             << "' identified. Function's alloc and dealloc functions erased "
+                "from memory functions.\n";
+      MemoryFunctions.erase(fnName);
     }
-  }
 
-  if (fnName == "main") {
-    functionIsKnown = true;
+    if (fnName == Pair.second) {
+      errs() << "**ANALYSIS-WARNING**: Re-definition of deallocation function '"
+             << fnName
+             << "' identified. Function's alloc and dealloc functions erased "
+                "from memory functions.\n";
+      MemoryFunctions.erase(fnName);
+    }
   }
 
-  if (!functionIsKnown && !ALLOW_REDEFINE) {
-    errs() << "**ANALYSIS-WARNING**: Unknown function '" << fnName
-           << "' identified and will be labelled as unsafe.\n";
-    UnsafeFunctions.insert(fnName);
-  }
+  if (fnName != "main") {
+    // annotation reasoning here; not complete yet but some of the reasoning is
+    // here
+
+    // plan: have a class wherein you pass it the function and it extracts the
+    // annotations on the function and the annotations on each parameter .need
+    // to first look at every paramater and assign {}. then analyze note:
+    // annotations on the return type will appear exactly the same as if they
+    // were on the function itself note: with LLVM IR, if you put two
+    // annotations on a var, theyll appear as two 'separate' names; e.g.,
+    // %param3.addr3 and %param3.addr4 refer to the same var. im assuming that
+    // if we use the same alias reasoning this'll be resolved easily
+
+    GlobalVariable *glob =
+        F.getParent()->getGlobalVariable("llvm.global.annotations");
+
+    std::string annotation;
+
+    for (llvm::Argument &arg : F.args()) {
+      logout("arg " << arg.getName().str())
+    }
+
+    if (glob != NULL) {
+      auto *ca = dyn_cast<ConstantArray>(glob->getInitializer());
+      for (unsigned i = 0; i < ca->getNumOperands(); ++i) {
+        if (ConstantStruct *structAn =
+                dyn_cast<ConstantStruct>(ca->getOperand(i))) {
+          if (ConstantExpr *expr =
+                  dyn_cast<ConstantExpr>(structAn->getOperand(0))) {
+            if (ConstantExpr *note =
+                    cast<ConstantExpr>(structAn->getOperand(1))) {
+              if (GlobalVariable *annotateStr =
+                      dyn_cast<GlobalVariable>(note->getOperand(0))) {
+                if (ConstantDataSequential *data =
+                        dyn_cast<ConstantDataSequential>(
+                            annotateStr->getInitializer())) {
+                  if (expr->getOpcode() != Instruction::BitCast ||
+                      expr->getOperand(0) != &F) {
+                    continue;
+                  }
+                  logout("annotation = " << data->getAsString().str())
+                }
```

Comment:
This code makes me very nervous, because it looks like you're fully deconstructing the annotations. That means that this code is going to be brittle to any changes in the program's structure or the annotation language; maybe that's okay for now.

Regardless, though, this code should be in its own helper function.

Reply:
Can I move this to a helper function in a different pull request wherein I further build annotation handling?

Reply:
Putting it into a helper function now is probably better, since we know we'll eventually want to do so.